pyOCD/gdbserver/gdbserver.py

```diff
@@ -329,12 +338,27 @@ def run(self):
                     break
 
                 if self.packet_io.interrupt_event.isSet():
-                    logging.debug("Got unexpected ctrl-c, ignoring")
+                    if self.non_stop:
+                        self.target.halt()
+                        self.is_target_running = False
+                        self.sendStopNotification()
+                    else:
+                        logging.debug("Got unexpected ctrl-c, ignoring")
                     self.packet_io.interrupt_event.clear()
 
+                if self.non_stop and self.is_target_running:
+                    try:
+                        if self.target.getState() == TARGET_HALTED:
+                            logging.debug("state halted")
+                            self.is_target_running = False
+                            self.sendStopNotification()
+                    except Exception as e:
+                        logging.debug("Unexpected exception: %s", e)
+                        traceback.print_exc()
+
                 # read command
                 try:
-                    packet = self.packet_io.receive()
+                    packet = self.packet_io.receive(block=not self.non_stop)
```

Comment:
Oh gotcha, that makes sense.  For some reason I was thinking that the reception of any data on the socket, including ctrl-c, would cause receive to return.  Now that you mention it, it is obvious why this is not the case.  I think the way you have it now is good.